File: node_project.py
import subprocess
import os
import json


def execute():
    return 1


def addDevDependencies():
    dependencies = ["@jest/globals", "jest", "nodemon", "ts-jest", "typescript"]
    npm_install = ["npm", "i", "-D"]
    install_command = npm_install
    for dependency in dependencies:
        install_command.append(" ")
        install_command.append(dependency)

    subprocess.run(install_command, cwd=os.getcwd())


def addScripts():
    loaded_json = None
    with open("package.json", "r") as package_json:
        loaded_json = json.load(package_json)

    with open("package.json", "w") as package_json:
        scripts = {
            "test": "jest unit",
            "integration": "jest int",
            "dev": "nodemon",
            "prod": "node build/src/main.js",
            "build": "tsc",
        }
        loaded_json["scripts"] = scripts
        json.dump(loaded_json, package_json)


def setupTsConfig():
    with open("tsconfig.json", "w") as tsconfig_json:
        tsconfig = {
            "compilerOptions": {"outDir": "build/src"},
            "exclude": ["**/*.test.ts"],
        }
        json.dump(tsconfig, tsconfig_json)


def setupGitIgnore():
    ignored_list = ["node_modules", "build"]
    with open(".gitignore", "w") as gitignore:
        ignored_string = ""
        for ignored in ignored_list:
            ignored_string = ignored_string + ignored + "\n"
        gitignore.write(ignored_string)


def setupJestConfig():
    config = """/** @type {import('ts-jest').JestConfigWithTsJest} */
module.exports = {
    preset: 'ts-jest',
    testEnvironment: 'node',
};"""
    with open("jest.config.js", "w") as jestConfig:
        jestConfig.write(config)


if __name__ == "__main__":
    addDevDependencies()
    addScripts()
    setupTsConfig()
    setupGitIgnore()
    setupJestConfig()

File: node_project/nodeProject.go
package nodeProject

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"strings"
)

func main() {
	fmt.Println("Hello, World")
}

func initWorkingDirectory(dirName string) {
	os.Mkdir(dirName, 0755)
	os.Chdir(dirName)
}

func addDevDependencies() {
	dependencies := []string{"@jest/globals", "jest", "nodemon", "ts-jest", "typescript"}
	for _, element := range dependencies {
		npmCommand := exec.Command("npm", "i", "-D", element)
		npmCommand.Output()
	}
}

func addScripts() {
	packageFile, _ := os.ReadFile("package.json")
	var packageJson map[string]interface{}
	json.Unmarshal(packageFile, &packageJson)
	scripts := map[string]string{
		"test":        "jest unit",
		"integration": "jest int",
		"dev":         "nodemon",
		"prod":        "node build/src/main.js",
		"build":       "tsc",
	}
	packageJson["scripts"] = scripts
	marshaledPackageJson, _ := json.Marshal(packageJson)
	os.WriteFile("package.json", marshaledPackageJson, 0644)
}

func setupTsConfig() {

	compileOptions := map[string]string{
		"outDir": "build/src",
	}

	exclude := []string{"**/*.test.ts"}

	tsConfigJson := make(map[string]interface{})

	tsConfigJson["compileOptions"] = compileOptions

	tsConfigJson["exclude"] = exclude

	marshaledTsConfigJson, _ := json.Marshal(tsConfigJson)

	os.WriteFile("tsconfig.json", marshaledTsConfigJson, 0644)

}

func setupGitIgnore() {
	ignored := []string{"node_modules", "build"}

	var sb strings.Builder

	for _, element := range ignored {
		sb.WriteString(element)
		sb.WriteString("\n")
	}

	os.WriteFile(".gitignore", []byte(sb.String()), 0644)
}

func setupJestConfig() {

	jestConfig := `/** @type {import('ts-jest').JestConfigWithTsJest} */
module.exports = {
    preset: 'ts-jest',
    testEnvironment: 'node',
};`
	os.WriteFile("jest.config.js", []byte(jestConfig), 0644)
}
